pkg/apis/mysql/v1alpha1/mysqluser_types.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2020 Pressinfra SRL
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NOTE: json tags are required. Any new fields you add must have json tags for
+// the fields to be serialized.
+
+// SecretKeySelector contains a key of the secret to select from.
+type SecretKeySelector struct {
+	// The name of the secret in the pod's namespace to select from.
+	corev1.LocalObjectReference `json:",inline"`
+	// The key of the secret to select from.  Must be a valid secret key.
+	Key string `json:"key"`
+}
+
+// MySQLUserSpec defines the desired state of MySQLUserSpec
+type MySQLUserSpec struct {
+	// ClusterRef represents a reference to the MySQL cluster.
+	// This field should be immutable.
+	ClusterRef corev1.ObjectReference `json:"clusterRef"`
+	// User is the name of the user that will be created with will access the specified database.
+	// This field should be immutable.
+	User string `json:"user"`
+	// Password is the password for the user.
+	Password SecretKeySelector `json:"password"`
+	// AllowedHost is the allowed host to connect from.
+	// This field should be immutable.
+	AllowedHost string `json:"allowedHost,omitempty"`
+	// Permissions is the list of roles that user has in the specified database.
+	Permissions []MySQLPermission `json:"permissions,omitempty"`
+
+	// AccountResourceLimits allow settings limit per mysql user as defined here:
+	// https://dev.mysql.com/doc/refman/5.7/en/user-resources.html
+	// +optional
+	AccountResourceLimits AccountResourceLimits `json:"accountResourceLimits,omitempty"`
```

Comment:
```suggestion
	ResourceLimits  corev1.ResourceList `json:"resourceLimits,omitempty"`
```

Reply:
I'm not sure that ResourceList from corev1 is the right choice here. I can define another one that is similar to that.

For example:
```go
type AccountResourceOptions map[AccountResourceOptionName]int

const (
  MaxUserConnections AccountResourceOptionName = "MAX_USER_CONNECTIONS"
  ...
)
```

Reply:
Using `corev1.ResourceList` is more Kubernetes-ish since you can define
```
spec:
  accountLimits:
    maxUserConnections: 1k
    maxUpdatesPerHour: 2M
```

Reply:
Is there any advantage of having our own `AccountResourceLimits` type? We can easliy have

```go
const(
  ResourceMaxUserConnections corev1.ResourceName = "MAX_USER_CONNECTIONS"
  ResourceMaxQueriesPerHour corev1.ResourceName = "MAX_QUERIES_PER_HOUR"
  ResourceMaxUpdatesPerHour corev1.ResourceName = "MAX_UPDATES_PER_HOUR"
  ResourceMaxConnectionsPerHour corev1.ResourceName = " MAX_CONNECTIONS_PER_HOUR"
)

```

Reply:
Ok, done.